### PR TITLE
msg/async, v2: add final auth message to authenticate pre-auth traffic

### PR DESCRIPTION
**Pull request by rzarzynski**
This PR bases on top of the https://github.com/ceph/ceph/pull/26865 clean-up. However, the area of merge conflict is quite small the clean-up commits could be dissected if necessary.

The work related to pre-auth is conveyed in last 5 commits.

### CODE REVIEW

**Comment by liewegas** [comment]
lots of crashes like
```
2019-03-09T23:51:11.451 INFO:tasks.ceph.mon.a.smithi168.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.1.0-528-g6e4f448/rpm/el7/BUILD/ceph-14.1.0-528-g6e4f448/src/auth/Crypto.h: In function 'sha256_digest_t CryptoKey::hmac_sha256(CephContext*, const bufferlist&)' thread 7f327b3ed700 time 2019-03-09 23:51:11.454063
2019-03-09T23:51:11.451 INFO:tasks.ceph.mon.a.smithi168.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.1.0-528-g6e4f448/rpm/el7/BUILD/ceph-14.1.0-528-g6e4f448/src/auth/Crypto.h: 175: FAILED ceph_assert(ckh)
2019-03-09T23:51:11.451 INFO:teuthology.orchestra.run.smithi168.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.1.0-528-g6e4f448/rpm/el7/BUILD/ceph-14.1.0-528-g6e4f448/src/auth/Crypto.h: In function 'sha256_digest_t CryptoKey::hmac_sha256(CephContext*, const bufferlist&)' thread 7fec9ad74700 time 2019-03-09 23:51:11.454088
2019-03-09T23:51:11.452 INFO:teuthology.orchestra.run.smithi168.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/14.1.0-528-g6e4f448/rpm/el7/BUILD/ceph-14.1.0-528-g6e4f448/src/auth/Crypto.h: 175: FAILED ceph_assert(ckh)
2019-03-09T23:51:11.452 INFO:tasks.ceph.mon.a.smithi168.stderr: ceph version 14.1.0-528-g6e4f448 (6e4f44841fa08929ec727e376a4d8afa1eb7db05) nautilus (rc)
2019-03-09T23:51:11.452 INFO:tasks.ceph.mon.a.smithi168.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x14a) [0x7f328b39a4b6]
2019-03-09T23:51:11.453 INFO:tasks.ceph.mon.a.smithi168.stderr: 2: (ceph::__ceph_assertf_fail(char const*, char const*, int, char const*, char const*, ...)+0) [0x7f328b39a684]
2019-03-09T23:51:11.453 INFO:tasks.ceph.mon.a.smithi168.stderr: 3: (ProtocolV2::finish_auth()+0x2d8) [0x7f328b673d28]
2019-03-09T23:51:11.453 INFO:tasks.ceph.mon.a.smithi168.stderr: 4: (ProtocolV2::run_continuation(Ct<ProtocolV2>&)+0x34) [0x7f328b6730e4]
2019-03-09T23:51:11.453 INFO:tasks.ceph.mon.a.smithi168.stderr: 5: (AsyncConnection::process()+0x186) [0x7f328b640ba6]
2019-03-09T23:51:11.453 INFO:tasks.ceph.mon.a.smithi168.stderr: 6: (EventCenter::process_events(unsigned int, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> >*)+0xa15) [0x7f328b694645]
2019-03-09T23:51:11.453 INFO:tasks.ceph.mon.a.smithi168.stderr: 7: (()+0x54d155) [0x7f328b699155]
2019-03-09T23:51:11.454 INFO:tasks.ceph.mon.a.smithi168.stderr: 8: (()+0x7dd31f) [0x7f328b92931f]
2019-03-09T23:51:11.454 INFO:tasks.ceph.mon.a.smithi168.stderr: 9: (()+0x7dd5) [0x7f3287f96dd5]
2019-03-09T23:51:11.454 INFO:tasks.ceph.mon.a.smithi168.stderr: 10: (clone()+0x6d) [0x7f3286e5cead]
```
http://pulpito.ceph.com/sage-2019-03-09_23:33:37-rados-wip-sage2-testing-2019-03-09-1325-distro-basic-smithi/

File: src/auth/Crypto.cc [hunk 1]
```diff
@@ -131,6 +131,20 @@ std::size_t CryptoKeyHandler::decrypt(
   return todo_len;
 }
 
+sha256_digest_t CryptoKeyHandler::hmac_sha256(
+  const ceph::bufferlist& in) const
+{
+  ceph::crypto::HMACSHA256 hmac((const unsigned char*)secret.c_str(), secret.length());
```
Comment (rzarzynski) [inline src/auth/Crypto.cc]: IIUC `ceph::crypto::HMACSHA256` still uses NSS inside. We should consider switching it to OpenSSL.

File: src/msg/async/ProtocolV2.cc [hunk 2]
```diff
@@ -94,6 +94,7 @@ ProtocolV2::ProtocolV2(AsyncConnection *connection)
       bannerExchangeCallback(nullptr),
       next_tag(static_cast<Tag>(0)),
       keepalive(false) {
+  ceph::crypto::init(cct);
```
Comment (rzarzynski) [inline src/msg/async/ProtocolV2.cc]: Just added this NSS-driven init. Currently we ship only `ceph::crypto::init` for NSS but it appears our builds have both `USE_NSS` and `USE_OPENSSL` defined while HMAC is only implemented with NSS. We will to change this, I guess.

**Comment by liewegas** [comment]
mons are having trouble talking: http://pulpito.ceph.com/sage-2019-03-12_02:13:11-rados-wip-sage2-testing-2019-03-11-1728-distro-basic-smithi/

File: src/common/ceph_crypto.h [hunk 3]
```diff
@@ -217,7 +199,114 @@ namespace ceph {
     };
   }
 }
+#endif
+
+#ifdef USE_OPENSSL
+namespace ceph::crypto::ssl {
+# if OPENSSL_VERSION_NUMBER < 0x10100000L
```
Comment (rzarzynski) [inline src/common/ceph_crypto.h]: This should make the Bionic and its OpenSSL version happy.

File: src/msg/async/ProtocolV2.cc [hunk 4]
```diff
@@ -224,6 +224,9 @@ void ProtocolV2::reset_recv_state() {
     auth_meta.reset(new AuthConnectionMeta);
     session_stream_handlers.tx.reset(nullptr);
     session_stream_handlers.rx.reset(nullptr);
+    pre_auth.enabled = true;
+    pre_auth.txbuf.clear();
+    pre_auth.rxbuf.clear();
```
Comment (rzarzynski) [inline src/msg/async/ProtocolV2.cc]: The `--ms_inject_socket_failures` showed that the cleaning on e.g. `_fault()` was missing.

**Comment by liewegas** [comment]
looks good: http://pulpito.ceph.com/sage-2019-03-13_17:48:35-rados-wip-sage2-testing-2019-03-13-0947-distro-basic-smithi/
the upgrade failure looks like http://tracker.ceph.com/issues/38345 (unrelated)

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

**Pull request merged**